File: Remove Duplicates From Sorted Array.py
# Link: https://leetcode.com/problems/remove-duplicates-from-sorted-array/submissions/

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        
        # Base cases: if list is empty or only has 1 element
        if nums == None or len(nums) == 0:
            return 0
        
        if len(nums) == 1:
            return 1
        
        # Create index variable
        indx = 1
        
        # Iterate list
        for i in range(len(nums) - 1):
            
            # Check if both numbers doesn't match
            if nums[i] != nums[i+1]:
              
                # Replace 1st old repeative number with next unique number
                # E.g. [1, 1, 2] => [1, 2, 2]
                nums[indx] = nums[i + 1]
                
                # Update index variable to next position
                # E.g. Index was 1 before this iteration based on previous example. Now, it is 2.
                indx += 1
                
        # Return last index, which is also the new size of the list
        return indx

File: Remove Duplicates from Sorted Array.java
// https://leetcode.com/problems/remove-duplicates-from-sorted-array/
// https://youtu.be/DEJAZBq0FDA?si=1TR4bryklCx_JKwV
/*
Time Complexity: O(n) where n = length of nums.

Space Complexity: O(1)
*/

class Solution {
    public int removeDuplicates(int[] nums) {

        // Return array's length if it has 0 or 1 elements
        if (nums.length <= 1) {
            return nums.length;
        }

        // Create a left pointer
        // IMPORTANT: I set leftRight = rightPtr because the very 1st element in nums will always be unique
        int leftPtr = 1;
        for (int rightPtr = leftPtr; rightPtr < nums.length; rightPtr++) {
            
            // If the last 2 elements are the same, skip to next iteration
            if (nums[rightPtr] == nums[rightPtr - 1]) {
                continue;
            }

            // If the last 2 elements are different, move new element to leftPtr and update leftPtr
            nums[leftPtr] = nums[rightPtr];
            leftPtr++;
        }

        // The first leftPtr elements are unique
        return leftPtr;
    }
}
